yjit/src/codegen.rs

```diff
@@ -4910,6 +4935,36 @@ fn gen_send_cfunc(
     EndBlock
 }
 
+/// Generate code to read the length of an array
+fn get_array_len(asm: &mut Assembler, array_reg: Opnd) -> Opnd
+{
+    asm.comment("get array length for embedded or heap");
+
+    // Pull out the embed flag to check if it's an embedded array.
+    let flags_opnd = Opnd::mem(VALUE_BITS, array_reg, RUBY_OFFSET_RBASIC_FLAGS);
+
+    // Get the length of the array
+    let emb_len_opnd = asm.and(flags_opnd, (RARRAY_EMBED_LEN_MASK as u64).into());
+    let emb_len_opnd = asm.rshift(emb_len_opnd, (RARRAY_EMBED_LEN_SHIFT as u64).into());
+
+    // Conditionally move the length of the heap array
+    let flags_opnd = Opnd::mem(VALUE_BITS, array_reg, RUBY_OFFSET_RBASIC_FLAGS);
+    asm.test(flags_opnd, (RARRAY_EMBED_FLAG as u64).into());
+
+    // Need to repeat this here to deal with register allocation
+    //let array_opnd = ctx.stack_opnd(0);
+    //let array_reg = asm.load(array_opnd);
```

Comment:
I'm experiencing a similar issue. I think the `array_reg` argument should be `array_opnd` and it should have the following code before the first `flags_opnd` and before `array_len_opnd`:

```rs
let array_reg = match array_opnd {
    Opnd::Reg(_) => array_opnd,
    _ => asm.load(array_opnd),
};
```